Motor claw

Lucimoot opened the pull request.

Lucimoot committed:
Added Intake subsystem

added intake subsystem for motors rather than pneumatics, added grab and release commands along with a basic safe state.

Lucimoot committed:
Used Experimental Logged Subsystem

I was Matt and Ben's Guinea Pig for an hour and a half :>
(according to my captors I was learning)

Lucimoot committed:
Game-Object specific grab/release commands, Color Sensor command

+ Cone grab/release commands
+ Cube grab/release commands
+ Color sensor grab/release command
- generic motor grab/release commands

bot190 commented on src/main/java/frc/robot/subsystems/Intake.java:
```diff
@@ -0,0 +1,82 @@
+
+public class Intake extends LoggedSubsystem<IntakeData, IntakeData.Map> {
+
+    private IDSolenoid solenoid = new MockDSolenoid();
```
We should be passing a `Value` object in the IntakeData object instead of creating the object here. (Yes I told you to do this, yes you do need to change it).

bot190 commented on src/main/java/frc/robot/subsystems/Intake.java:
```diff
@@ -0,0 +1,82 @@
+            if (DetectedColor == Color.kBlue) {
+                cmd().onInitialize(() -> {
+                    getData().setPoint = GRAB_SPEED;
+                }).runsUntil(() -> DetectedColor != Color.kBlue).onEnd(() -> {
```
Why are we stopping the intake when the color changes? Shouldn't we be looking at the distance or something?

bot190 commented on src/main/java/frc/robot/subsystems/Intake.java:
```diff
@@ -0,0 +1,82 @@
+    private IDSolenoid solenoid = new MockDSolenoid();
+    private final double GRAB_SPEED = 1;
+    private final double RELEASE_SPEED = -1;
+    private IColorSensor colorSensor;
```
This also needs to be moved into the map. and the IntakeData should probably have a double for the distance and a color value.

bot190 reviewed (changes requested):
The big thing is the sensor and solenoid need to be moved into the `IntakeMap`, and be a part of the `IntakeData` object.

Lucimoot committed:
Moved Variables to Map + Logged Variables

/ cone grab/release commands
/ cube grab/release commands
/ color grab/release commands
+ added all variables needs for edited commands to the map
+ logged all added map variables

msoucy commented on src/main/java/frc/robot/maps/RobotMap.java:
```diff
@@ -7,14 +7,22 @@
 
     // $Getters$
+    public IntakeData.Map getIntakeMap() {
+        return intakeMap;
```
Can you merge line 17 with this? So it's just `return new IntakeData.Map()`?

msoucy commented on src/main/java/frc/robot/subsystems/Intake.java:
```diff
@@ -0,0 +1,84 @@
+    @Override
+    public void periodic() {
+        super.periodic();
+        sensorColor = new Color(getData().detectedColor[0], getData().detectedColor[1], getData().detectedColor[2]);
```
You should be doing this conversion in `IntakeData` instead

msoucy commented on src/main/java/frc/robot/maps/subsystems/IntakeData.java:
```diff
@@ -0,0 +1,68 @@
+            this.colorSensor = new MockColorSensor();
+        }
+
+        public Map(SmartMotorController motor) {
```
Why are there three different constructors?

Lucimoot committed:
Reworked Color Sensor command and added distance checks

/ color senor command
+ gamePieceDistance var
+ minGamePieceDistance var
+ maxGamePieceDistance var
+ Logging for gamePieceDistance

Liammurray19 committed:
added variables in intakemap

Lucimoot committed:
saved intake data changes and properly pushed it

i made an oopsie and this is just fixing it (sorry liam and lucas)

msoucy committed:
Overhaul sensor-based control function
